### PR TITLE
Generalize CartesianGrid for multiple dimensions and variable cell lengths

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:13]:
```suggestion
#ifndef ARBORX_DETAILS_CARTESIAN_GRID_HPP
#define ARBORX_DETAILS_CARTESIAN_GRID_HPP
```

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsFDBSCANDenseBox.hpp:191]: Why did you template that guy?

**Comment by dalg24** [comment]
This is false advertising.  You are doing more than moving things around.  You did changes to the class.  Please provide a description.

**Review by dalg24 — COMMENTED** [review]
Incomplete review.  Will continue after the description is updated

**Comment by aprokop** [comment]
> This is false advertising. You are doing more than moving things around. You did changes to the class. Please provide a description.

Sorry, did not notice it took the title from the first commit.

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsFDBSCANDenseBox.hpp:191]: It takes `CartesianGrid` as an argument.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:27]: Why templating on `Box` rather than the number of dimensions?

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:35]: Why are these public?

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:34]: Make sure the description of the PR captures that change

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:20]: What do you use from that header?

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:27]: We want to be able to return a box corresponding to a cell index. If we just provide dimension, it is unclear what box type is and it won't work with both old and new geometries.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:35]: Mostly out of convenience so that we don't have to provide access operators. We print out the sizes in DBSCAN, which is a very useful information.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:20]: Yes, it's unnecessary.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (masterleinad) [inline src/details/ArborX_DetailsCartesianGrid.hpp:87]: I really dislike these implicit `bool` conversions.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (masterleinad) [reply src/details/ArborX_DetailsCartesianGrid.hpp:30]: I would prefer a getter function in that case.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (masterleinad) [inline src/details/ArborX_DetailsCartesianGrid.hpp:57]:
```suggestion
      int i = Kokkos::Experimental::floor((point[d] - min_corner[d]) / _h[d]);
```

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (masterleinad) [inline src/details/ArborX_DetailsCartesianGrid.hpp:96]: This can only happen in the first branch.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [reply src/details/ArborX_DetailsCartesianGrid.hpp:23]: I still think we should template the grid over the dimension, not the box type

Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:23]: Well, this PR would then have to wait until we merge all others, because `cellBox` will have to return the experimental box, which right now cannot be used in DBSCAN.

**Comment by aprokop** [comment]
@dalg24 OK, I templated the `CartesianGrid` on the dimension instead of the `Box`. The PR had to be expanded to also include DBSCAN working with different Box types (new and old geometries). But everything works now. Please re-review

**Comment by aprokop** [comment]
I'll need to run some DBSCAN experiments to make sure no performance penalty occurred.

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp
Comment (dalg24) [reply src/details/ArborX_DetailsFDBSCANDenseBox.hpp:191]: My point is you can deduce the type entirely.
Primitives -> DIM -> CartesianGrid

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:90]:
```suggestion
  auto extent(int d) const { assert(0 <= d && d < DIM);  return _n[d]; }
```

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (dalg24) [reply src/details/ArborX_DetailsCartesianGrid.hpp:67]: ping

File: src/details/ArborX_MinimumSpanningTree.hpp
Comment (dalg24) [inline src/details/ArborX_MinimumSpanningTree.hpp:592]: Now that I see how many times you have that code I almost want to define a `Box<DIM>` for pre 3.7 Kokkos

File: examples/dbscan/ArborX_DBSCANVerification.hpp
Comment (aprokop) [reply examples/dbscan/ArborX_DBSCANVerification.hpp:304]: I'll update #729 after we merge this.

File: src/ArborX_DBSCAN.hpp
Comment (aprokop) [reply src/ArborX_DBSCAN.hpp:58]: Hyper-sphere does not have a constructor that takes in `double`. Original sphere does.

File: src/ArborX_DBSCAN.hpp
Comment (aprokop) [reply src/ArborX_DBSCAN.hpp:102]: Later.

File: src/ArborX_DBSCAN.hpp
Comment (aprokop) [reply src/ArborX_DBSCAN.hpp:106]: We could be strict in our access traits validation to make sure we return only ArborX point and box types.

File: src/ArborX_DBSCAN.hpp
Comment (aprokop) [reply src/ArborX_DBSCAN.hpp:183]: Yes. I'm not sure why, but clang-format changed few things around the code. Both using VIM plugin and the check format script.

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:67]: Are you suggesting adding `KokkosExt::floor`?

File: src/details/ArborX_DetailsCartesianGrid.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:90]: Sure.

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsFDBSCANDenseBox.hpp:191]: That's fair.

File: src/details/ArborX_MinimumSpanningTree.hpp
Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]: It would help some, but would still have some clutter as long as we have to maintain backwards compatibility with older geometries. Hopefully, that's for a short time until Kokkos 4.0 release.

Comment (dalg24) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]:
```C++
namespace ExperimentalHyperGeometry {
template <int DIM>
struct Box : struct ArborX::Box {
  using AborX::Box::Box;
};
}

Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]: I'm not sure what you mean. We can't inherit as we need to use new point types as corners.

Comment (dalg24) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]: in the Kokkos pre-3.7 code path

Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]: Ah, OK. With a static assert `DIM == 3`? And I assume same with Point and Sphere. I can do that, but would you mind if I do that after we merge this PR?

Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:592]: PR was rebased on #734.

File: src/ArborX_DBSCAN.hpp
Comment (dalg24) [inline src/ArborX_DBSCAN.hpp:70]: I really dislike the amount/mix of template parameters here.
You can deduce the dimension and then the grid type from `PointPrimitives below`

File: src/ArborX_DBSCAN.hpp
Comment (dalg24) [inline src/ArborX_DBSCAN.hpp:393]: I don't like so much using directly `BasicBoundingVolumeHierarchy` but I don't have a better idea

File: src/ArborX_DBSCAN.hpp
Comment (aprokop) [reply src/ArborX_DBSCAN.hpp]: Done. It's ugly, but will be fixed with #729.

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
Confirmed no performance penalty running few DBSCAN and MST experiments using master (aca4aba5) and this branch (b23049cd) on OACISS Saturn (Nvidia A100).

Merging with HIP skipped (not started due to timeout).

**Pull request merged**
